inc/class-tachyon.php

```diff
@@ -956,4 +956,79 @@ public function cleanup_rest_image_downsize( $response ) {
 	public function _override_image_downsize_in_rest_edit_context() {
 		return true;
 	}
+
+	/**
+	 * Rename an existing attachment file to support Tachyon.
+	 *
+	 * Some legacy uploads may have image dimensions in the file name. Tachyon
+	 * does not support this for performance reasons. This function renames
+	 * attachments
+	 *
+	 * @param integer $attachment_id The attachment post ID.
+	 * @param bool $remove_old_files If true deletes the old files.
+	 * @return true|WP_Error
+	 */
+	public static function _rename_file( int $attachment_id, bool $remove_old_files = true ) {
+		global $wpdb;
+
+		$file = get_attached_file( $attachment_id );
+
+		// Trim dimensions from file name and make sure it's actually different.
+		$new_file = preg_replace( '/-\d+x\d+\.(jpe?g|png|gif)$/', '.$1', $file );
+		if ( $file === $new_file ) {
+			return new WP_Error( 'tachyon_rename_file', "{$file} does not need to be renamed" );
+		}
+
+		// Make sure it's unique.
+		$new_file = wp_unique_filename( dirname( $file ), basename( $new_file ) );
+		$new_file = dirname( $file ) . DIRECTORY_SEPARATOR . $new_file;
+
+		// Copy old file to new name.
+		$copied = copy( $file, $new_file );
+		if ( ! $copied ) {
+			return new WP_Error( 'tachyon_rename_file', "{$file} could not be copied to {$new_file}" );
+		}
+
+		// Generate new metadata.
+		$metadata = wp_generate_attachment_metadata( $attachment_id, $new_file );
+		if ( empty( $metadata ) || ! is_array( $metadata ) ) {
+			return new WP_Error( 'tachyon_rename_file', "Could not generate new attachment metadata for attachment ID {$attachment_id}" );
+		}
+
+		// Update the attachment.
+		wp_update_attachment_metadata( $attachment_id, $metadata );
+		update_attached_file( $attachment_id, $new_file );
+		clean_attachment_cache( $attachment_id );
+
+		// Update posts table content.
+		$wpdb->query( 'SET @uids := 0;' );
+		$wpdb->query( $wpdb->prepare(
+			"UPDATE {$wpdb->posts}
+			SET post_content = REPLACE(post_content, %s, %s)
+			WHERE post_content LIKE %s
+				AND ( SELECT @uids := CONCAT_WS(',', ID, @uids) );",
+			_wp_relative_upload_path( $file ),
+			_wp_relative_upload_path( $new_file ),
```

Comment:
Do we need to do this replacement for all sizes? If we are removing all sizes, I'm assuming so.

Also, what is the subquery here doing?

Reply:
Damn, good point, we would need to check all sizes yes. Maybe it's worth looking again at the alternative approach of checking if the file exists :/

The sub query is keeping track of all modified post IDs to know which caches to update.

Reply:
Removed this approach in favour of WP CLI search & replace. Only sure fire method. Might be ok as you spin up separate processes using `WP_CLI::run_command()`.